dean.js

```diff
@@ -0,0 +1,136 @@
+/** @param {NS} ns **/
+/*
+create list of all servers
+iterate through list:
+count programs
+open ports
+nuke
+scp and exec hack script
+if all 5 ports are open then everything above has run, remove from list
+*/
+export async function main(ns) {
+	//create array of all servers except purchased called listServers[]
+	var listServers = await ns.scan('home');
+	let listIndex = 0;
+
+	while (listIndex < listServers.length) {
+		await ns.sleep(50);
+		let listScan = ns.scan(listServers[listIndex], true);
+		for (let i = 0; i < listScan.length; i++) {
+			if (listServers.indexOf(listScan[i]) === -1) {
+				listServers[listServers.length] = listScan[i];
+			}
+		}
+		//at this point the full list includes purchased servers
+		listIndex++;
+	}
+	//make array of purchased servers
+	var persServers = await ns.getPurchasedServers();
+	//if I have purchased servers, remove them from listServers
+	if (persServers.length > 0) {
+		for (let i = 0; i < persServers.length; i++) {
+			if (listServers.indexOf(persServers[i]) != -1) {
+				let spliceIndex = listServers.indexOf(persServers[i]);
+				listServers.splice(spliceIndex, 1);
+			}
+		}
+	}
```

Comment:
1) Do purchased servers not need to run `h-early`? 
2) I would maybe do this before the while loop so that the while loop iterates through less items
3) I would learn to use `Array.filter()` as its a bit less code, and easier to read IMO

```suggestion
  let persServers = await ns.getPurchasedServers();
  let listServers = await ns.scan('home').filter(host => !persServers.includes(host));
```

Reply:
this script just has the servers that can run h/w/g on themselves, as a way to get a a tiny extra bit of hacking experience early in bitnodes, purchased servers are filled by a different script that runs w/g 12 times a minute

Reply:
removed my loop and added your stuff at the top of main, need to wrap my brain around the filter stuff still especially "=>"